numba/unicode.py

```diff
@@ -375,6 +375,18 @@ def _find(substr, s):
     return -1
 
 
+@njit
+def _rfind(substr, s, begin, e):
+    if e == -1:
+        e = len(s)
+    i = e - len(substr)
+    while i >= begin:
+        if _cmp_region(s, i, substr, 0, len(substr)) == 0:
+            return i
+        i = i - 1
+    return -1
```

Comment:
I'm not sure this algorithm is correct. Consider:
```python
from numba import njit

@njit
def foo(src, sub, start=0, end=-1):
    return src.rfind(sub, start, end)

a = 'aabbccddbb'
b = 'bb'
s = -1
e = 8

x = foo.py_func(a, b, s, e)
y = foo(a, b, s, e)
print(x)
print(y)
assert x == y
```
this returns 2 when it should return -1.

Reply:
I will change the algorithm for such corner cases.

Reply:
Great, thanks.

Reply:
@stuartarchibald I am sorry for the confusion but I am a bit doubtful about this answer. Why should it return -1 instead of 2? rfind method return highest index for the substring inside string[start: end] then why does it return -1?

Reply:
Because:
```
In [10]: 'aabbccddbb'[-1: 8]                                                                                                    
Out[10]: ''
```
and `bb` is not in the empty string

Reply:
Oh Okay! Got it now! Thanks

Reply:
Great, thanks. I'd recommend playing around on the REPL and investigating a bit about what happens with various input combinations, these can become the basis of the unit tests to check this code.